### ISSUE TITLE
VLM thinking mode: model falls into repetition loop; entire reasoning buffer flushed into a single content delta when generation ends inside thinking (Qwen3.8-27B-nvfp4 + MTP drafter)

### ISSUE THREAD

**Issue by dalianblue** [jundot/omlx#3129]
## Environment

- oMLX 0.6.3rc3 (Homebrew tap), macOS 26.5, Apple M1 Pro / 32GB
- Model: `mlx-community/Qwen3.8-27B-nvfp4` (VLM, served via VLM engine) + `mlx-community/Qwen3.8-27B-MTP-nvfp4` drafter attached through per-model settings (`vlm_mtp_enabled: true`, `vlm_mtp_draft_model: ...`)
- Server: `omlx serve --model-dir ~/.omlx/models --paged-ssd-cache-dir ~/.omlx/cache --port 8000`, MTP acceptance ~98.8% (247/250, block_size 3) — speculative decoding itself works great

## Symptom 1: repetition loop in thinking mode (stochastic)

With thinking enabled (default chat template) and `temperature: 0.2`, some Chinese medical/knowledge questions send the model into an infinite self-questioning loop inside the reasoning channel until `max_tokens`. Example prompt:

```
告诉我2025年最新版《中国2型糖尿病防治指南》中，糖化血红蛋白（HbA1c）的控制目标具体修改为多少？
```

Observed: the phrase `让我仔细回忆：2024年1月，CDS发布了...` repeated **212×** over 622s until `max_tokens` (8192). The same prompt re-run later completed cleanly in 173s, so it is stochastic. A second prompt (brain-death criteria, ~700-word structured answer) looped on 3/3 runs.

Notably, adding `frequency_penalty: 0.3 + presence_penalty: 0.1` made things *worse* on some prompts: a TCA-cycle question that answered correctly without penalties went into a loop **2/2 runs** with penalties (874s / 871s). I later noticed `model_settings.py` documents that vlm_mtp is exclusive with processor-backed penalty settings — but the OpenAI-compatible API silently accepts `frequency_penalty`/`presence_penalty` instead of rejecting them, which makes this unsupported combination easy to hit by accident.

## Symptom 2: entire reasoning buffer flushed into a single `content` delta (deterministic given symptom 1)

Whenever generation terminates while still inside the thinking section, the full accumulated reasoning text is emitted as **one giant `content` delta**, duplicating the whole thinking into the message content.

Delta-level analysis of one run (M1 + penalties, 871s):

```
total deltas: 4066 | content-only: 1 | reasoning-only: 4062 | dual-field: 0
content chars: 23774 | reasoning chars: 23774 | content == reasoning: True
```

The single content-only delta contains all 23,774 characters of reasoning, byte-identical to the concatenated `reasoning_content` stream. So the visible "answer" becomes a wall of raw thinking text.

This flush also occurs in the default, penalty-free configuration: two questions in a 57-question benchmark run ended with `content` starting with thinking-style prose (`用户询问：“...”首先，我需要检索...`) instead of the actual answer.

## Reproduction

```bash
curl -s http://127.0.0.1:8000/v1/chat/completions \
  -H 'Content-Type: application/json' \
  -d '{"model":"Qwen3.8-27B-nvfp4","messages":[{"role":"user","content":"请写出三羧酸循环（TCA循环）中，唯一一次底物水平磷酸化反应，并注明催化酶和生成物。禁止使用模糊词，给出具体数值或机制。"}],"temperature":0.2,"max_tokens":8192,"stream":true,"frequency_penalty":0.3,"presence_penalty":0.1}'
```

Count deltas carrying only `content` vs only `reasoning_content`; on a looping run the concatenations come out identical (one flush delta at the end). Without penalties the loop is stochastic — retry a few times or use a long structured-answer prompt (e.g. brain-death vs vegetative state criteria).

Relevant server log lines from a looping run:

```
vlm_mtp decode started: request=... 
vlm_mtp stats: request=... finish=stop rounds=125 accepted=247/250 (98.8%) tokens_per_round=2.98 emitted=372 block_size=3
```

## Expected behavior

1. If generation ends inside the thinking section, the reasoning text should stay in the `reasoning_content` channel (or be dropped), not be flushed into `content`.
2. Ideally the API should reject (or at least warn about) penalty/processor-backed parameters when `vlm_mtp` is enabled, since the combination currently degrades sampling in a hard-to-diagnose way.

Happy to provide the full 57-question benchmark transcript and delta logs if useful. Thanks for a great project — the MTP acceptance rates and prefix caching are genuinely impressive.

**Comment by dalianblue** [comment]
**Correction + additional log evidence** after digging through `~/.omlx/logs/server.log`:

The penalty-related degradation is *not* an unsupported combination being executed — oMLX detects per-request logits processors and **silently falls back from vlm_mtp to the plain BatchGenerator path**, logging only an INFO line:

```
2026-08-25 13:31:42 - omlx.scheduler - INFO - vlm_mtp routing skipped for 1c37bb99-...: request carries per-request logits processors without vlm_mtp support (function, function); falling back to BatchGenerator
2026-08-25 14:00:27 - omlx.scheduler - INFO - vlm_mtp routing skipped for bd18c4cf-...: (same)
2026-08-25 14:15:14 - omlx.scheduler - INFO - vlm_mtp routing skipped for 40377ad0-...: (same)
2026-08-25 14:20:13 - omlx.scheduler - INFO - vlm_mtp routing skipped for c3185eb4-...: (same)
2026-08-25 14:50:51 - omlx.scheduler - INFO - vlm_mtp routing skipped for 33cccee2-...: (same)
```

These 5 lines correspond 1:1 to the five penalty-carrying requests in my testing (the looping M1 runs included). So:

1. The repetition loops observed **with** penalties happened on the fallback (non-MTP) path.
2. The loops observed **without** penalties (the ×212 run, and the brain-death prompt 3/3) went through the normal vlm_mtp path — so the loop is not specific to either decode path.
3. For completeness, the vlm_mtp stats of the clean re-run of the HbA1c prompt (no penalties, 171s, correct refusal-style answer):

```
vlm_mtp stats: finish=stop rounds=822 accepted=1030/1644 (62.7%) tokens_per_round=2.25 emitted=1853 block_size=3
```

vs. short clean answers which typically show 85–100% acceptance.

Suggestion (replacing my earlier "reject the parameters" ask): it would help if the fallback were visible to the client — e.g. a warning log level, a response header, or a note in the response — since sending `frequency_penalty` currently disables speculative decoding with no signal at all, which is easy to misdiagnose as a general perf/quality regression (it took me a benchmark round-trip to notice).

Symptom 2 (whole reasoning buffer flushed into a single `content` delta when generation ends inside thinking) stands unchanged — it occurred on both paths.
